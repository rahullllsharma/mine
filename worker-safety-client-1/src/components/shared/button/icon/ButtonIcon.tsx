import { IconSpinner } from "@/components/iconSpinner";
import type { IconName } from "@urbint/silica";
import { Icon } from "@urbint/silica";
import cx from "classnames";
import type { ButtonHTMLAttributes, ForwardedRef } from "react";
import { forwardRef } from "react";

export type ButtonIconProps = Omit<
  ButtonHTMLAttributes<HTMLButtonElement>,
  "children"
> & {
  iconName: IconName;
  loading?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
};

function ButtonIcon(
  { iconName, className, loading = false, ...props }: ButtonIconProps,
  ref: ForwardedRef<HTMLButtonElement>
): JSX.Element {
  return (
    <button
      {...props}
      disabled={props.disabled || loading}
      ref={ref}
      className={cx(
        "text-xl text-neutral-shade-75 disabled:text-neutral-shade-38 disabled:cursor-not-allowed",
        className
      )}
    >
      {loading ? <IconSpinner className="flex" /> : <Icon name={iconName} />}
    </button>
  );
}

export default forwardRef(ButtonIcon);
