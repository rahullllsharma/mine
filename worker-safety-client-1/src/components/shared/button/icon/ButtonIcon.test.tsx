import { fireEvent, render, screen } from "@testing-library/react";
import ButtonIcon from "./ButtonIcon";

describe("ButtonIcon", () => {
  describe("when is rendered", () => {
    it("should match the snapshot", () => {
      const { asFragment } = render(<ButtonIcon iconName="settings_filled" />);
      expect(asFragment()).toMatchSnapshot();
    });

    it("should be in the document", () => {
      render(<ButtonIcon iconName="settings_filled" />);
      expect(screen.getByRole("button")).toBeInTheDocument();
    });
  });

  describe("when is loading", () => {
    it("should be disabled and showing the loading icon", () => {
      const { asFragment } = render(
        <ButtonIcon loading iconName="settings_filled" />
      );
      expect(screen.getByRole("button")).toBeDisabled();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when is disabled", () => {
    it('should have the "disabled" attribute', () => {
      render(<ButtonIcon iconName="settings_filled" disabled />);
      expect(screen.getByRole("button")).toBeDisabled();
    });
  });

  describe("when is clicked", () => {
    const mockOnClick = jest.fn();

    it('should have the "onClick" callback to be called', () => {
      render(<ButtonIcon iconName="settings_filled" onClick={mockOnClick} />);
      fireEvent.click(screen.getByRole("button"));
      expect(mockOnClick).toBeCalled();
    });
  });
});
