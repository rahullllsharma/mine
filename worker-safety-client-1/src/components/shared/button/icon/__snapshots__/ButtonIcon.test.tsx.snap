// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonIcon when is loading should be disabled and showing the loading icon 1`] = `
<DocumentFragment>
  <button
    class="text-xl text-neutral-shade-75 disabled:text-neutral-shade-38 disabled:cursor-not-allowed"
    disabled=""
  >
    <i
      aria-hidden="true"
      class="animate-spin flex"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.75 17.25c-.615.615-1.626.622-2.125-.091a9 9 0 01-.446-.706c-.43-.756.008-1.666.828-1.958l.263-.095c.674-.24 1.404.123 1.786.727.382.605.397 1.42-.108 1.925l-.198.198zM10.144 4.81c-.217-.842.29-1.715 1.158-1.783A9 9 0 118.848 20.43c-.814-.305-1.061-1.285-.62-2.034.442-.75 1.408-.975 2.248-.748a5.85 5.85 0 101.621-11.497c-.87-.015-1.735-.498-1.953-1.34z"
        />
      </svg>
    </i>
  </button>
</DocumentFragment>
`;

exports[`ButtonIcon when is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="text-xl text-neutral-shade-75 disabled:text-neutral-shade-38 disabled:cursor-not-allowed"
  >
    <i
      aria-hidden="true"
      class="ci-settings_filled"
    />
  </button>
</DocumentFragment>
`;
